importer/src/main/java/org/hiero/mirror/importer/downloader/block/BlockNode.java

```
@@ -0,0 +1,266 @@
+                            serverSuccess = true;
+                            break;
+                        }
+
+                        throw new BlockStreamException("Received status " + response.getStatus() + " from block node");
+                    }
+                    default -> throw new BlockStreamException("Unknown response case " + response.getResponseCase());
+                }
+
+                errors.set(0);
+
+                if (serverSuccess) {
+                    break;
+                }
+            }
+        } catch (BlockStreamException ex) {
+            onError();
+            throw ex;
+        } catch (Exception ex) {
+            onError();
+            throw new BlockStreamException(ex);
+        } finally {
+            if (grpcCall.get() != null) {
+                grpcCall.get().cancel("unsubscribe", null);
+            }
+        }
+    }
+
+    @Override
+    public int compareTo(BlockNode other) {
+        return COMPARATOR.compare(this, other);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("BlockNode(%s)", properties.getEndpoint());
+    }
+
+    public BlockNode tryReadmit(boolean force) {
+        if (!active && (force || Instant.now().isAfter(readmitTime.get()))) {
+            active = true;
+        }
+
+        return this;
+    }
+
+    private void onError() {
+        if (errors.incrementAndGet() >= streamProperties.getMaxSubscribeAttempts()) {
+            active = false;
+            errors.set(0);
+            readmitTime.set(Instant.now().plus(streamProperties.getReadmitDelay()));
+            log.warn(
+                    "Marking connection to {} as inactive after {} attempts",
+                    this,
+                    streamProperties.getMaxSubscribeAttempts());
+        }
+    }
+
+    private class BlockAssembler {
+
+        private long loadStart;
+        private final List<List<BlockItem>> pending = new ArrayList<>();
+        private int pendingCount = 0;
```

Why can't we flatten the block items into a single dimenson list when we receive them? The fact that their aggregated during send is only for performance reasons upstream. Then we can ditch pendingCount as well.

---

performance consideration on our end

- avoid flattening when the whole block is in a single stream response
- defend ourselves against malicious block node which tries to stream a huge block (many block items), so we don't waste cycles to keep flattening them.

---

Disagree here. We protect against such a scenario by limiting the stream response size. If we also need to protected against too many block items in a response then we should throw an exception up front if it exceeds that threshold. I think it will be cleander code and slightly faster performance for happy path since we don't have to allocate a separate array and flatten at the end.

---

The performance difference is negligible. And for a good behaving block node, it's actually the opposite because a good block node should stream a complete block in a single stream response, the logic here would avoid the unnecessary flattening.